invenio_records_resources/services/files/transfer.py

```diff
@@ -0,0 +1,157 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio-Records-Resources is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Files transfer."""
+
+from abc import ABC, abstractmethod
+from enum import Enum
+
+from invenio_db import db
+from invenio_files_rest.errors import FileSizeError
+
+
+class TransferType(str, Enum):
+    """File type, it inherits from str to be JSON serializable.
+
+    LOCAL represents a file that is stored locally in the instance's storage.
+    EXTERNAL represents a file that needs to be fetched from an external storage
+             and saved locally.
+    LINKED represents a file that is stored externally and is linked to the record.
+    """
+
+    LOCAL = "S"  # S = Stored
+    EXTERNAL = "E"
+    LINKED = "L"
+
+    def __eq__(self, other):
+        """Equality test."""
+        return self.value == other
+
+    def __str__(self):
+        """Return its value."""
+        return self.value
+
+
+class BaseTransfer(ABC):
+    """Local transfer."""
+
+    def __init__(self, type):
+        """Constructor."""
+        self.type = type
+
+    @abstractmethod
+    def init_file(self, record, file_metadata):
+        """Initialize a file."""
+        pass
+
+    def set_file_content(self, record, file, file_key, stream, content_length):
```

Comment:
TODO: create generator to disallow set by users when file==external, allowing only system identity (tasks)